File: solutions/typescript/grade-school/1/grade-school.ts
declare global {
  interface ObjectConstructor {
    groupBy<T, K extends PropertyKey>(
      items: Iterable<T>,
      callback: (item: T) => K
    ): Record<K, T[]>
  }
}

export class GradeSchool {
  private school: Map<string, number> = new Map()
  
  roster(): Record<number, string[]> {
    return Object.fromEntries(
      Object.entries(
        Object
          .groupBy(this.school.entries(), ([_, grade]) => grade))
          .map(([grade, students]) => [grade, students.map(([name, _]) => name).sort()])
    )
  }

  add(student: string, grade: number): void {
    this.school.set(student, grade)
  }

  grade(level: number): string[] {
    return this.roster()[level] || []
  }
}

File: solutions/python/grade-school/1/grade_school.py
class School:
    def __init__(self):
        self._name_grade = {}
        self._added = []

    def add_student(self, name, grade):
        added = name not in self._name_grade
        if added:
            self._name_grade[name] = grade
        self._added.append(added)

    def roster(self):
        return [
            name for name, _ in sorted(
                self._name_grade.items(),
                key = lambda item: (item[1], item[0])
            )
        ]

    def grade(self, grade_number):
        return [
            name for name, grade in sorted(
                self._name_grade.items(),
                key = lambda item: item[0]
            ) if grade == grade_number
        ]

    def added(self):
        return [added for added in self._added]
